# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/painting/nsDisplayList.cpp
# Commit: ae6b08b3f4de
# Full Hash: ae6b08b3f4def1cf7cf12b8cfa53d0edaad99f97
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2022-02-23 09:38:00
# Regressor Bug: 1714584
# File Overlap Count: 1
# Description:
#   Bug 1714584 - Part 2: Remove nsDisplayList::RemoveBottom() r=mstange
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D138153
# ==============================================================================

diff -r a261bb633790 -r ae6b08b3f4de layout/painting/nsDisplayList.cpp
--- a/layout/painting/nsDisplayList.cpp	Tue Feb 22 23:42:18 2022 +0000
+++ b/layout/painting/nsDisplayList.cpp	Tue Feb 22 23:42:18 2022 +0000
@@ -2326,8 +2326,7 @@
 }
 
 void nsDisplayList::DeleteAll(nsDisplayListBuilder* aBuilder) {
-  nsDisplayItem* item;
-  while ((item = RemoveBottom()) != nullptr) {
+  for (auto* item : TakeItems()) {
     item->Destroy(aBuilder);
   }
 }
@@ -4674,17 +4673,14 @@
 static nsresult WrapEachDisplayItem(nsDisplayListBuilder* aBuilder,
                                     nsDisplayList* aList,
                                     nsDisplayItemWrapper* aWrapper) {
-  nsDisplayList newList;
-  nsDisplayItem* item;
-  while ((item = aList->RemoveBottom())) {
+  for (nsDisplayItem* item : aList->TakeItems()) {
     item = aWrapper->WrapItem(aBuilder, item);
     if (!item) {
       return NS_ERROR_OUT_OF_MEMORY;
     }
-    newList.AppendToTop(item);
+    aList->AppendToTop(item);
   }
   // aList was emptied
-  aList->AppendToTop(&newList);
   return NS_OK;
 }
 
@@ -8543,15 +8539,14 @@
   // 1,2: backgrounds and borders
   aOutResultList->AppendToTop(BorderBackground());
   // 3: negative z-index children.
-  for (;;) {
-    nsDisplayItem* item = PositionedDescendants()->GetBottom();
-    if (item && item->ZIndex() < 0) {
-      PositionedDescendants()->RemoveBottom();
+  for (auto* item : PositionedDescendants()->TakeItems()) {
+    if (item->ZIndex() < 0) {
       aOutResultList->AppendToTop(item);
-      continue;
-    }
-    break;
-  }
+    } else {
+      PositionedDescendants()->AppendToTop(item);
+    }
+  }
+
   // 4: block backgrounds
   aOutResultList->AppendToTop(BlockBorderBackgrounds());
   // 5: floats
